### PR TITLE
Feature: alt-click visibility eye to toggle show just that layer (hide others)

### PR DESCRIPTION
**Pull request by psobolewskiPhD**
# Description

This PR implements an alt/option-click on the layer visibility icon (eye), that mimics Photoshop:
- On Alt/Option-click, that singular layer is shown and everything else is hidden. 
- If you alt-click the *same layer* again, then the visibility state of all layers will be restored. 
- If you alt-click a different layer, that one will be shown and everything else will be hidden.

To achieve this, the visibility state of layers is stored when toggled.

Overall, this feature is super handy for isolating a single layer but not losing the overall visibility of other layers in the case of a complex stack.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update: please see docs PR: https://github.com/napari/docs/pull/121

# References
Partially addresses: https://github.com/napari/napari/issues/5253 and https://github.com/napari/napari/issues/1338
 - it doesn't use double click, because I think that could be error prone, instead alt/option-click (like Photoshop), which is more of a deliberate decision.

cc: @haesleinhuepf @GenevieveBuckley

Partially addresses https://github.com/napari/napari/issues/2550
 - it doesn't change the colormap

Note: this was split off, as feature creep, from https://github.com/napari/napari/pull/5574
 
# How has this been tested?

- [x] example: added a test to ensure the alt-click logic works properly
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation: https://github.com/napari/docs/pull/121
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/_qt/containers/_layer_delegate.py
Comment (Czaki) [inline napari/_qt/containers/_layer_delegate.py]: hm. maybe we should use `is` instead of `==`. As far as Know it will be same result in this case but will be more correct in the context of semantics.

File: napari/_qt/containers/_layer_delegate.py
Comment (Czaki) [inline napari/_qt/containers/_layer_delegate.py]: same here

**Review by Czaki — COMMENTED** [review]
The _show_on_alt_click_hide_others looks like something that could be  tested

File: napari/_qt/containers/_layer_delegate.py
Comment (GenevieveBuckley) [reply napari/_qt/containers/_layer_delegate.py]: Agree that `layer.visible = layer is clicked_layer` might be a little easier to read and understand here. I don't feel super strongly about it, though.

**Review by GenevieveBuckley — DISMISSED** [review]
I tried it out and this feature feels very nice to use. I like it a lot!

Discoverability will be the main problem for users here, so I'd like there to be a corresponding PR adding this information to the docs. The [napari viewer tutorial page](https://napari.org/stable/tutorials/fundamentals/viewer.html) might be a good place for that.

**Comment by psobolewskiPhD** [comment]
> I tried it out and this feature feels very nice to use. I like it a lot!
> 
> Discoverability will be the main problem for users here, so I'd like there to be a corresponding PR adding this information to the docs. The [napari viewer tutorial page](https://napari.org/stable/tutorials/fundamentals/viewer.html) might be a good place for that.

Yup, definitely will add to the docs for this just as soon as the updated tutorial is merged:
https://github.com/napari/docs/pull/65

**Comment by psobolewskiPhD** [comment]
> The _show_on_alt_click_hide_others looks like something that could be tested

Went down a long rabbit hole trying to get qtbot to alt click on the visibility icon.
Gave up.
But with the way this method is defined, it turns out the behavior is easy to test by just calling it on an index, given a model.
I bet that's why @Czaki suggested that in the first place and the above note was supposed to be a hint 😅 
Wasted a lot of time, but learned a bit in the process.
Anyhow, I think everything should be set now—docs PR pending merge of [docs/#65](https://github.com/napari/docs/pull/65) as noted above.

**Comment by psobolewskiPhD** [comment]
Here's the promised docs PR: https://github.com/napari/docs/pull/121
Please see:
https://output.circle-artifacts.com/output/job/86a6f901-d644-4b99-b644-daa5bf643dfb/artifacts/0/docs/_build/tutorials/fundamentals/viewer.html#layer-list
and
https://output.circle-artifacts.com/output/job/86a6f901-d644-4b99-b644-daa5bf643dfb/artifacts/0/docs/_build/guides/layers.html#layer-visibility

**Review by brisvag — COMMENTED** [review]
Did you make sure what happens if layers get added/removed between "show only" and "restore visibility"?

**Comment by Czaki** [comment]
> Did you make sure what happens if layers get added/removed between "show only" and "restore visibility"?

The new layer state will be not touched.

**Comment by psobolewskiPhD** [comment]
@brisvag 
> Did you make sure what happens if layers get added/removed between "show only" and "restore visibility"?

Yes, that logic is here (not sure why I can't link to the changed code directly... it's line 237 and on):
```
elif self._alt_click_layer() is clicked_layer:
             # second alt-click on same layer, so restore visibility
             # account for any added/deleted layers when restoring
             for layer in layer_list:
                 if layer in self._layer_visibility_states:
                     layer.visible = self._layer_visibility_states[layer]
```
If a layer was deleted, it's not in `layer_list` so everything is fine.
If a layer was added, it's not in the stored states list (`self._layer_visibility_states`), so it's visibility is not affected.

Do you want me to try to add these cases to the test?
I think I can `append` and `remove` from `LayerList`, right?

**Comment by brisvag** [comment]
> I think I can `append` and `remove` from `LayerList`, right?

Yup. Yes please!

**Comment by psobolewskiPhD** [comment]
> > I think I can `append` and `remove` from `LayerList`, right?
> 
> Yup. Yes please!

@brisvag Done! ✅

**Comment by brisvag** [comment]
Sorry, one more thing: what happens (or should happen) if I shift click, then change visibility of another layer, then shift click again?

**Comment by psobolewskiPhD** [comment]
> Sorry, one more thing: what happens (or should happen) if I shift click, then change visibility of another layer, then shift click again?

You mean alt-click right? 
If you alt-click to show just one layer (say layer 1), you can normal click to show a second layer (say layer 3). So you will have two showing for comparison or such. 
If you then alt-click again on layer 1 it will restore the visibility state as it was prior to the alt-click-show-just-layer-1. Layer 3 will return to that state as well, so if it was originally hidden, it will re-hide.
If instead you alt-click on some other layer, say layer 5, it will hide layer 1 and 3 and show just layer 5. A second-alt-click on layer 5 will restore the state prior to the **original alt-click**.

So the alt-click mode starts on first alt-click and ends when a layer is alt-clicked a *second time* in a row.

Edit: thinking more, there is the possible case of use alt-click to show just layer 1 (at time = 0), then manually showing some other layers, doing some work, yada yada, over time and then alt-clicking on layer 1 again, just to show it alone. In this specific case, the view state from before the first alt-click (from time = 0) will be restored.

**Comment by brisvag** [comment]
Ok, thanks for clarifying. Not sure if there's a good/sensible way to handle this, since it's a bit tricky, but it's fine by me. Maybe clarify this in the docs as well if possible!

File: napari/_qt/containers/_tests/test_qt_layer_list.py
Comment (Czaki) [inline napari/_qt/containers/_tests/test_qt_layer_list.py:73]: As I understand this also checks the synchronization of `layer.visible` to the checkbox state?

Maybe it is obsolete and we could make the test shorter?

**Review by Czaki — DISMISSED** [review]

File: napari/_qt/containers/_tests/test_qt_layer_list.py
Comment (psobolewskiPhD) [reply napari/_qt/containers/_tests/test_qt_layer_list.py:73]: You mean because the two tests above `test_set_layer_invisible_makes_item_unchecked` and `test_set_item_unchecked_makes_layer_invisible` already ensure there will be no disconnect between states?

**Comment by psobolewskiPhD** [comment]
> Edit: thinking more, there is the possible case of use alt-click to show just layer 1 (at time = 0), then manually showing some other layers, doing some work, yada yada, over time and then alt-clicking on layer 1 again, just to show it alone. In this specific case, the view state from before the first alt-click (from time = 0) will be restored.

Yeah, so this is a real use case and kinda annoying.
If you alt-click a layer to isolate it, then manually show another layer, do some stuff, and later want to alt-click to isolate a layer you will get the layer state from before the first alt-click, e even though it's not relevant.
I had my wife check in Photoshop:
- by chaining alt-clicks, you can cycle through isolated layers and 2nd alt-click restores the state from before alt-clicking --  this works like in this PR right now
- if you regular click to show a layer after alt-clicking (one ore more times), the stored visibility state is reset, so the state-to-be-restored is updated/set on every regular click. This is not how the current PR works, but is more logical. If you manually set the visibility, then it means you want those to be restored later.

**Comment by psobolewskiPhD** [comment]
OK I think I have the logic sorted now.
In `_layer_delagate` I now catch regular clicks—but pass them on so they still work normally—to save the visibility state and clear any alt-click states.
This way alt-click to isolate followed by regular clicks to define specific visibility resets the state again, but chained alt-clicks cycle the solo layers.
On my end the behavior mimics Photoshop.

File: napari/_qt/containers/_layer_delegate.py
Comment (psobolewskiPhD) [inline napari/_qt/containers/_layer_delegate.py]:
```suggestion
        """When toggling visibility of a layer, store the current visibility state of
```

**Comment by psobolewskiPhD** [comment]
I'm going to dismiss the reviews, because while the behavior is similar, the actual implementation is almost completely different from what was approved.

**Comment by jni** [comment]
@psobolewskiPhD looking at the implementation, personally I would like to move much of it to the LayerList model. That is, alt-click on a button should call a method/function `toggle_single_layer_visible(layerlist, layer)` (or similar) that then goes and does the thing. In my opinion, GUI code should be as thin as possible and defer as much logic as possible to model-side code. But maybe this is very GUI specific and I'm being annoying? I just found the Qt code hard to follow and I kinda sorta think that shouldn't be the case... 😅

**Comment by Czaki** [comment]
@jni for me, it is in a good place. From one side it is really GUI specific way to interact. I do not think taha anyone may want to perform this via API. 

Also, it will require to write many lines of additional code to mimic qt structures.

**Comment by psobolewskiPhD** [comment]
We discussed this at PR party.
Leaning towards keeping it a GUI/Qt implementation, but I will verify how it interacts with programatic toggling of visibility or selection state.
I'm putting it on my to-do!

**Review by jni — APPROVED** [review]
All good! Definitely don't want to block. Thanks @psobolewskiPhD this is going to be super handy! 😃 🚀

**Comment by psobolewskiPhD** [comment]
I've simplified things even more! Now programatic change of visibility (or adding a layers) is accounted for properly—IMO. 
No matter how the visibility is changed, alt-click stores the current visibility state and shows the single layer. 2nd alt-click on same layer restores the state from before first alt-click.
Regular clicks reset the alt-click state, so we're back to normal.
Consecutive alt-clicks don't alter the saved state, just show single layers.

The only question is whether programmatic toggle of layer visibility should also reset alt-click state. I think no? Because it doesn't make sense to reset a GUI state without a GUI interaction. But I can be convinced otherwise I guess
Upside is programmatic toggle doesn't do anything extra, which is something @kephale was concerned about during the PR party.

**Comment by psobolewskiPhD** [comment]
@Czaki If you get some free time, can you sanity check this final implementation? No rush though, it's not urgent at all!
❤️

**Review by Czaki — APPROVED** [review]
It will be nice to have 100% test coverage on  `_show_on_alt_click_hide_others`. I have no other remarks.

**Comment by GenevieveBuckley** [comment]
This one looks ready to merge. All CI checks are passing, two reviewers have approved, the last reviewer comments were addressed in [21b0b64](https://github.com/napari/napari/pull/5618/commits/21b0b643b0afa18c84f46c93cf98297ab1fc6d30) and nothing else left outstanding.

Thanks @psobolewskiPhD for all the hard work, and to Juan and Grzegorz for their detailed reviews

**Pull request merged**
